Handle lost sparseness in non http data sources

akalenyu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Handle some non-http sources where we lose sparseness similarly to https://github.com/kubevirt/containerized-data-importer/pull/3219

For this to be complete, we must fix the broken check for sparseness in e2e, to help us test against regression.
The first issue is that we look at content size since #2168,
which is not beneficial in the sparseness check as opposed to disk usage.

The second issue is that the check we have in tests for sparseness is not honest because of the "equal to" part.
The virtual size has to be strictly greater than the size on disk.

The third issue is that we almost always resize which results in significantly larger virtual size.
What we really care about is comparing against the original virtual size of the image.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

akalenyu committed:
Correct wrong check of resulting image sparsiness

The first issue is that we look at content size since 2168,
which is not beneficial in the sparseness check as opposed to disk usage.

The second issue is that the check we have in tests for sparsiness is not honest because of the "equal to" part.
The virtual size has to be strictly greater than the content
(as the content is displayed by tools that understand sparseness).

The third issue is that we almost always resize which results in significantly larger virtual size.
What we really care about is comparing against the original virtual size of the image.

Signed-off-by: Alex Kalenyuk <akalenyu@redhat.com>

akalenyu commented:
/test all

akalenyu committed:
Use scratch space for non-archive uploads

Basically a follow up to 3219 for upload sources,
which suffer from the same issue of losing sparseness.

Signed-off-by: Alex Kalenyuk <akalenyu@redhat.com>

akalenyu commented:
/test pull-containerized-data-importer-fossa

alromeros commented on pkg/importer/upload-datasource.go:
```diff
@@ -51,8 +54,7 @@ func (ud *UploadDataSource) Info() (ProcessingPhase, error) {
 	}
-	if !ud.readers.Convert {
-		// Uploading a raw file, we can write that directly to the target.
+	if ud.directWriteException {
```
Maybe pass the source content type and do the `sourceContentType == common.BlockdeviceClone` here? It makes this easier to debug and understand than having an arbitrary `directWriteException` that's not too informative. Just a nit though, feel free to ignore.

akalenyu replied:
yeah that makes sense

alromeros reviewed (commented):
Really nice, just a couple of comments

akalenyu commented:
/hold
- vddk - no actual bug, fake vddk plugin lying
- imageio https://github.com/kubevirt/containerized-data-importer/blob/dd518e770bfc4131aec497386b2920c4a7d8cec9/pkg/importer/imageio-datasource.go#L119-L121

mhenriks commented on tests/framework/pvc.go:
```diff
@@ -284,24 +283,15 @@ func (f *Framework) VerifyBlankDisk(namespace *k8sv1.Namespace, pvc *k8sv1.Persi
 
 // VerifySparse checks a disk image being sparse after creation/resize.
-func (f *Framework) VerifySparse(namespace *k8sv1.Namespace, pvc *k8sv1.PersistentVolumeClaim, imagePath string) (bool, error) {
+func (f *Framework) VerifySparse(namespace *k8sv1.Namespace, pvc *k8sv1.PersistentVolumeClaim, imagePath string, originalVirtualSize int64) (bool, error) {
```
Why not get rid of this function and check `!VerifyImagePreallocated()`?

akalenyu replied:
Mainly because of the third point in the PR description

> The third issue is that we almost always resize which results in significantly larger virtual size.
What we really care about is comparing against the original virtual size of the image.

We could get rid of one of them by negating the other but have to adapt for that

mhenriks replied:
> What we really care about is comparing against the original virtual size of the image.

Why?

akalenyu replied:
Since current virtual size is post-resize you always end up comparing 1Gi or more (requested disk image size) versus a modest couple of megabytes, which is always going to pass as sort of a false negative

mhenriks commented on tests/framework/pvc.go:
```diff
@@ -284,24 +283,15 @@ func (f *Framework) VerifyBlankDisk(namespace *k8sv1.Namespace, pvc *k8sv1.Persi
-	return info.VirtualSize >= imageContentSize, nil
+	fmt.Fprintf(ginkgo.GinkgoWriter, "INFO: VerifySparse comparison: OriginalVirtual: %d vs SizeOnDisk: %d\n", originalVirtualSize, info.ActualSize)
+	// The size on disk of a sparse image is significantly lower than the image's virtual size
+	return originalVirtualSize-info.ActualSize >= units.MB, nil
```
What is the 1MB fudge factor all about?

awels replied:
Its probably to ensure we align the size on 1Mib. qemu gets unhappy if the image is not aligned on the actual storage block size.

mhenriks replied:
I think this check is a little too strict, wondering why current virtual size is not appropriate

akalenyu replied:
So I looked up the 1MiB->1MB switch.
It's because `qemu-img convert` only manages to shave off one mega**byte** for our test image
```INFO: VerifySparse comparison: OriginalVirtual: 18874368 vs SizeOnDisk: 17829888```

mhenriks replied:
So this check is specific to a particular image?

akalenyu replied:
No, you pass the virtual size of the used image as a parameter

akalenyu committed:
tests: skip sparse check for imageio/vddk

- ImageIO raw images are not passed through any sparsification mechanism
- VDDK ones do, but the fake plugin used for testing lies that they're
  not

Signed-off-by: Alex Kalenyuk <akalenyu@redhat.com>

akalenyu commented:
/retest

alromeros commented:
/remove-lifecycle rotten

akalenyu commented:
/reopen

akalenyu commented:
/hold cancel
/cc @mhenriks @alromeros @noamasu 
I revived this, not sure how nobody complains about lost sparseness

akalenyu commented:
/remove-lifecycle rotten

akalenyu commented:
Sparseness broken on all HPP lanes, hmm, weird

akalenyu commented:
> Sparseness broken on all HPP lanes, hmm, weird

apparently that image isn't sparse enough to test on
